### PR TITLE
fix(servicecatalogappregistry): Associate an application with attribute group

### PR DESCRIPTION
**Pull request by santanugho**
`Application-AttributeGroup` association happens in `ApplicationStack`. 
Therefore before the deployment of `ApplicationStack`, `AttributeGroup` stack should have been deployed. 
But with `ApplicationAssociator `where we associate all the stacks with the application (created as a part of `ApplicationAssociator`), attributeGroup stack now depends upon `ApplicationAssociator` stack to be created (since ResourceAssociation happens inside ResourceStack).
This creates a circular dependency and hence cdk synth fails. We found this bug during internal testing
This PR address this circular dependency issue, by allowing the customers of `ApplicationAssociator` to associate an attribute group in attribute group stack.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by TheRealAmazonKendra — DISMISSED** [review]
Thank you for your contribution, we cannot provide a meaningful review until there is a passing build. If you have questions about the build failure, please let us know.

Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request)). Specifically, a fix should state the bug, not the solution.

Lastly, is there an open issue for this? If so, please link it in the PR body.

**Comment by TheRealAmazonKendra** [comment]
Thanks for getting the build passing. The last two pieces of my comment will still need to be addressed.

> Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request)). Specifically, a fix should state the bug, not the solution.

> Lastly, is there an open issue for this? If so, please link it in the PR body.

**Comment by santanugho** [comment]
> Thanks for getting the build passing. The last two pieces of my comment will still need to be addressed.
> 
> > Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request)). Specifically, a fix should state the bug, not the solution.
> 
> > Lastly, is there an open issue for this? If so, please link it in the PR body.

PR title confirms to the conventional commit standard. This PR is for a bug fix in `servicecatalogappregistry` and the PR title is as per that standard.
This bug has been reported internally and there is no github issue open for it.

File: packages/@aws-cdk/aws-servicecatalogappregistry/README.md [hunk 1]
```diff
@@ -99,6 +99,45 @@ const associatedApp = new appreg.ApplicationAssociator(app, 'AssociatedApplicati
...
+    super(scope, id, props);
+    const myAttributeGroup = new appreg.AttributeGroup(app, 'MyFirstAttributeGroup', {
+      attributeGroupName: 'MyAttributeGroupName',
+      description: 'Test attribute group',
+      attributes: {},
+    });
+
+    this.attributeGroup = myAttributeGroup;
+  }
+}
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-servicecatalogappregistry/README.md:121]: Do we really expect people to define a stack for their AttributeGroup? Couldn't it go into one of the other stacks?

For example, wouldn't it make sense for a user to do:

```ts
const associatedApp = new appreg.ApplicationAssociator(app, 'AssociatedApplication', {
  applications: [ /* ... */],
  attributes: { 
    // ...
  },
});
```

File: packages/@aws-cdk/aws-servicecatalogappregistry/README.md [hunk 2]
```diff
@@ -191,6 +230,16 @@ declare const attributeGroup: appreg.AttributeGroup;
...
 ```
 
+### Associating an attribute group with application
+
+You can associate an application with an attribute group with `associateApplicationWithAttributeGroup`:
+
+```ts
+declare const application: appreg.Application;
+declare const attributeGroup: appreg.AttributeGroup;
+attributeGroup.associateApplicationWithAttributeGroup(application);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-servicecatalogappregistry/README.md:240]: This is good, although the method name is a little long. Given that the subject is of type `AttributeGroup` and the object is of type `Application`, do you really think we need the words "Application" and "AttributeGroup" in the method name? 

Why not

```ts
attributeGroup.associate(application);
// or
attributeGroup.associateWith(application);
// or
attributeGroup.associateTo(application);
```

File: packages/@aws-cdk/aws-servicecatalogappregistry/README.md (hunk 1 above)
Comment (santanugho) [reply packages/@aws-cdk/aws-servicecatalogappregistry/README.md:121]: `attributes` is not a valid input as per [ApplicationAssociatorProps](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-servicecatalogappregistry/lib/application-associator.ts#L10). Attribute group can definitely be a part of some other stack, however customers need to explicitly associate the attribute group to the application, as `ApplicationAssociator` is responsible to associate the stacks within the scope, it wont automatically associate the underlying AGs within a stack.

File: packages/@aws-cdk/aws-servicecatalogappregistry/README.md (hunk 2 above)
Comment (santanugho) [reply packages/@aws-cdk/aws-servicecatalogappregistry/README.md]: Yeah `associateWith` makes a lot more sense. Changed the same in the latest version.

File: packages/@aws-cdk/aws-servicecatalogappregistry/README.md (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-servicecatalogappregistry/README.md:121]: > attributes is not a valid input as per ApplicationAssociatorProps

I understand. 

The nice thing about software is that it is not like the laws of physics. It is in fact changeable. 

* FACT: "attributes" can not be passed to the ApplicationAssociator
* QUESTION: would it be better/easier/more convenient/more obvious IF "attributes" could be passed to ApplicationAssociator?

**Review by rix0rrr — CHANGES_REQUESTED** [review]
See my previous comment: do we want to force the user to create an attributegroup themselves?

I think users would prefer to minimize the amount of stacks in their account if possible.

File: packages/@aws-cdk/aws-servicecatalogappregistry/README.md (hunk 1 above)
Comment (santanugho) [reply packages/@aws-cdk/aws-servicecatalogappregistry/README.md:121]: TBH I like this idea of including `attribute-group` in `application-associator` construct. However, can we first prioritize this bug fix which will unblock our customers and they should be able to associate an attribute group to an application.

I will do some more deep dive on including `attribute-group` in `application-associator`, its pros and cons and will let you know about our plan to modify `application-associator` in order to include `attribute-group`.

Let me know if this plan works for you.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**